### PR TITLE
Add single file range read feature

### PR DESCRIPTION
**Pull request by davidfarkas**
### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.md

### CODE REVIEW

**Comment by nagem** [comment]
I also added @kofalt as a reviewer. For more info: this is functionality that we'll need sooner than the Object Storage PR can land, which is why I asked David to pull it into it's own PR.

**Comment by kofalt** [comment]
Given that this does not use a library, and instead hand-implements the range requests, my suggestion would be to toggle this behavior via a URL parameter that is intended only to be used for viewers.

```python
if not 'http header, viewer-range=1':
    # old behavior that is currently deleted
else:
    # new behavior added as part of this PR
```

That way, the range-request stuff only fires for the viewer code that requires it, and does not fire for ordinary usage such as downloads. This ensures clients (browsers, CLI, python, etc) don't interact with the logic and lets us ship this as-is without having to extensively validate the implementation.

**Comment by ambrussimon** [comment]
@kofalt The goal was to implement this for object storage, specifically to provide the same (standard) interface as any storage provider does, _even when using local storage_. Effort went into testing it with both local and google storage, and verifying that it's working the same way (or better). It was separated into it's own PR as per @nagem's request to not have two implementations lurking around on different branches.

I don't think it should go into a URL param, as the goal is to solve it for good in a standard fashion. As you said, effort is needed for validation though.

PS: Not using a lib was a design choice (avoiding dep creep), but for one I'm easily convinced to use 3rd party implementation.

File: tests/unit_tests/python/test_util.py [hunk 1]
```diff
@@ -1,5 +1,39 @@
+import pytest
+
 from api import util
 
+@pytest.fixture(scope='function', params=[
+    #range header content     expected_output
+    ('bytes=1-5',       [(1, 5)]),
```
Comment (ambrussimon) [inline tests/unit_tests/python/test_util.py]: Please clean up the whitespace here.

File: tests/unit_tests/python/test_util.py [hunk 2]
```diff
@@ -1,5 +1,39 @@
...
+    ('bytes=15, -6--10, 10-15',   util.ParseError),
+    ('bytes=1-5; 6-10; 10-15',   util.ParseError),
+])
+def parse_range_header_fixture(request):
+    header, expected_output = request.param
+    return header, expected_output
+
+
+def test_parse_range_header(parse_range_header_fixture):
+    input, expected_output = parse_range_header_fixture
```
Comment (ambrussimon) [inline tests/unit_tests/python/test_util.py]: `input` shadows a builtin here.

File: api/handlers/listhandler.py [hunk 3]
```diff
@@ -480,13 +480,66 @@ def get(self, cont_name, list_name, **kwargs):
...
+                                self.response.write('Content-Type: %s\n' % str(
+                                    fileinfo.get('mimetype', 'application/octet-stream')))
+                                self.response.write('Content-Range: %s' % 'bytes %s-%s/%s\n' % (str(first),
+                                                                                              str(last),
+                                                                                              str(fileinfo['size'])))
+                                self.response.write('\n')
+                                self.response.write(data)
+                                self.response.write('\n')
+                            else:
+                                self.response.headers['Content-Type'] = str(
```
Comment (ambrussimon) [inline api/handlers/listhandler.py]: I think setting any headers before the loop might be cleaner - even if it introduces a redundant `if len(ranges) > 1`.

File: api/util.py [hunk 4]
```diff
@@ -15,6 +15,9 @@
 from django.conf import settings
 from django.template import Template, Context
 
+BYTE_RANGE_REGEX = r'^(?P<first>\d+)-(?P<last>\d+)?$'
```
[the remark this answers is not in the document]
Comment (kofalt) [reply api/util.py]: Personally I was going to advocate for non-regex parsing, but it's not important is it's feature-flagged. Up to ya'll.

File: api/handlers/listhandler.py [hunk 5]
```diff
@@ -480,13 +480,66 @@ def get(self, cont_name, list_name, **kwargs):
...
+                                if last > fileinfo['size']:
+                                    length = fileinfo['size'] - first
+                                else:
+                                    length = last - first + 1
+
+                            f.seek(first, mode)
+                            data = f.read(length)
+
+                            if len(ranges) > 1:
+                                self.response.write('--THIS_STRING_SEPARATES\n')
```
[the remark this answers is not in the document]
Comment (kofalt) [reply api/handlers/listhandler.py]: Agreed on this part, with a quick skim I didn't see any `SHOULD`s / `MUST`s around using a unique string

One thing that I didn't see addressed in my (quick) review is that the response data stream needs to be escaped. Did I miss it?

File: api/handlers/listhandler.py [hunk 6]
```diff
@@ -480,13 +480,66 @@ def get(self, cont_name, list_name, **kwargs):
...
         else:
-            self.response.app_iter = open(filepath, 'rb')
-            self.response.headers['Content-Length'] = str(fileinfo['size']) # must be set after setting app_iter
-            if self.is_true('view'):
-                self.response.headers['Content-Type'] = str(fileinfo.get('mimetype', 'application/octet-stream'))
-            else:
-                self.response.headers['Content-Type'] = 'application/octet-stream'
-                self.response.headers['Content-Disposition'] = 'attachment; filename="' + filename + '"'
+
+            range_header = self.request.headers.get('Range', '')
```
[the remark this answers is not in the document]
Comment (kofalt) [reply api/handlers/listhandler.py]: I'm pretty sure they do, but I could be wrong...

File: api/handlers/listhandler.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidfarkas) [reply api/handlers/listhandler.py]: @kofalt escape how? I don't really understand.

File: api/handlers/listhandler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidfarkas) [reply api/handlers/listhandler.py]: Yep, webob capitalizes them. To make sure, I added a test, which sends lowercase range header.

**Comment by nagem** [comment]
Thanks for making the requested changes. LGTM to merge in it's adjusted state but I'll let @ambrussimon finalize his review as well.

**Comment by nagem** [comment]
Note: After being confused by the code returning the entire file with a 200 when unable to parse the range request, @ambrussimon  noted the `Note` in this section of the rcf: https://tools.ietf.org/html/rfc7233#section-4.4

Their testing with GCS showed that they also 200 and serve the entire file when encountering an improper range request.

File: api/handlers/listhandler.py [hunk 7]
```diff
@@ -480,13 +480,67 @@ def get(self, cont_name, list_name, **kwargs):
...
+
+                if self.is_true('view'):
+                    self.response.headers['Content-Type'] = str(fileinfo.get('mimetype', 'application/octet-stream'))
+                else:
+                    self.response.headers['Content-Type'] = 'application/octet-stream'
+                    self.response.headers['Content-Disposition'] = 'attachment; filename="' + filename + '"'
             else:
-                self.response.headers['Content-Type'] = 'application/octet-stream'
-                self.response.headers['Content-Disposition'] = 'attachment; filename="' + filename + '"'
+                self.response.status = 206
```
Comment (kofalt) [inline api/handlers/listhandler.py]: Are we sure `try - except - else` is a pattern we want to use here? It feels weird to grab the whole stack & throw `Feature flag not set` to fill the role of an if statement.

Comment (nagem) [reply api/handlers/listhandler.py]: We want to "fail back" to serving the whole file for any parsing issue. I think in this case we consider the feature flag to be part of our modified RFC7233 protocol. If we didn't make that assumption, we would still need an if statement outside a try/except block where the except executes the same code. That seems even more confusing.

**Pull request merged**